RFC: Enabling some commonly enabled compiler diagnostics

I've compiled a list of commonly enabled compiler diagnostics that we are currently choosing not to enable.

This is the list:

| Compiler diagnostic | no# of emitted unique GCC warnings in `master` | no# of emitted unique Clang warnings in `master` |
| ------------- | ------------- | ------------- |
| `-Wconditional-uninitialized`: Warn if a variable may be uninitialized when used  | Not supported | 2 |
| `-Wdouble-promotion`: Warn if `float` is implicit promoted to `double` | 1 | 8 |
| `-Wduplicated-branches`: Warn if `if`/`else` branches have duplicated code  | 0 | Not supported |
| `-Wduplicated-cond`: Warn if `if`/`else` chain has duplicated conditions  | 0 | Not supported |
| `-Wfloat-equal`: Warn if floating-point values are used in equality comparisons | 29 | 18 |
| `-Wlogical-op`: Warn about logical operations being used where bitwise were probably wanted  | 0 | Not supported |
| `-Wnon-virtual-dtor`: Warn the user if a class with virtual functions has a non-virtual destructor. This helps catch hard to track down memory errors.  | 22 | 10 |
| `-Wnull-dereference`: Warn if a potential `nullptr` dereference is detected  | 48 | Not supported |
| `-Woverloaded-virtual`: Warn if you overload (not `override`) a virtual function  | 0 | 0 |
| `-Wsuggest-override`: Warn about overriding virtual functions that are not marked with the `override` keyword | 303 (of which 192 are in `src/leveldb/` | Not supported |
| `-Wunreachable-code-loop-increment`: Warn if a loop will run only once (loop increment never executed)  | Not supported | 1 |
| `-Wunused-member-function`: Warn on unused member function  | Not supported | 2 |
| `-Wunused-template`: Warn on unused template | Not supported | 1 |

There is a large overlap between this list and [Jason Turner's list of recommended compiler diagnostics in the Collaborative Collection of C++ Best Practices (`cppbestpractices`) project](https://github.com/lefticus/cppbestpractices/blob/master/02-Use_the_Tools_Available.md#gcc--clang). There is also an overlap with the recommendations given in the [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines) (with editors Bjarne Stroustrup and Herb Sutter).

I'm now seeking feedback regarding these diagnostics: which of these would make sense to enable in the Bitcoin Core project? :)

> Any feedback? Should any of the suggested checks be removed before being submitted as a PR?

> -Wimplicit-fallthrough:

We explicitly disable this (just below), so having it in this diff doesn't enable anything. I also don't think we'd turn it on, given that the original reasons it was disabled still apply (large number of warnings in Boost, leveldb and tinyformat etc, see #10489). This is also part of `-Wextra` for GCC.

> -Wunused-template

This introduces a large number of Qt related warnings for me:
```bash
  CXX      qt/libbitcoinqt_a-platformstyle.o
In file included from qt/test/wallettests.cpp:1:
In file included from ./qt/test/wallettests.h:5:
In file included from /usr/local/Cellar/qt/5.14.0/lib/QtTest.framework/Headers/QTest:1:
In file included from /usr/local/Cellar/qt/5.14.0/lib/QtTest.framework/Headers/qtest.h:449:
In file included from /usr/local/Cellar/qt/5.14.0/lib/QtTest.framework/Headers/qtestsystem.h:45:
/usr/local/Cellar/qt/5.14.0/lib/QtCore.framework/Headers/qtestsupport_core.h:55:31: warning: unused function template 'qWaitFor' [-Wunused-template]
Q_REQUIRED_RESULT static bool qWaitFor(Functor predicate, int timeout = 5000)
                              ^
1 warning generated.
```

For everything below, ACK opening a PR, assuming you are fixing any warnings where applicable, and/or have submitted PRs upstream. Note however that this was only doing a quick check on macOS, with brew installed dependencies, so the output could be much worse on other platforms. I also did a quick check on Ubuntu Bionic and saw similar results.

> -Wdouble-promotion

```bash
torcontrol.cpp:702:51: warning: implicit conversion increases floating-point precision: 'float' to 'double' [-Wdouble-promotion]
    struct timeval time = MillisToTimeval(int64_t(reconnect_timeout * 1000.0));
                                                  ^~~~~~~~~~~~~~~~~ ~
1 warning generated.

qt/splashscreen.cpp:54:32: warning: implicit conversion increases floating-point precision: 'float' to 'qreal' (aka 'double') [-Wdouble-promotion]
    pixmap.setDevicePixelRatio(devicePixelRatio);
           ~~~~~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~
qt/splashscreen.cpp:60:61: warning: implicit conversion increases floating-point precision: 'float' to 'qreal' (aka 'double') [-Wdouble-promotion]
    QRadialGradient gradient(QPoint(0,0), splashSize.width()/devicePixelRatio);
                    ~~~~~~~~              ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
2 warnings generated.

qt/trafficgraphwidget.cpp:86:94: warning: implicit conversion increases floating-point precision: 'float' to 'double' [-Wdouble-promotion]
    painter.drawText(XMARGIN, YMARGIN + h - h * val / fMax-yMarginText, QString("%1 %2").arg(val).arg(units));
                                                                                         ~~~ ^~~
qt/trafficgraphwidget.cpp:96:98: warning: implicit conversion increases floating-point precision: 'float' to 'double' [-Wdouble-promotion]
        painter.drawText(XMARGIN, YMARGIN + h - h * val / fMax-yMarginText, QString("%1 %2").arg(val).arg(units));
                                                                                             ~~~ ^~~
2 warnings generated.
```

> -Wunreachable-code-loop-increment

```bash
init.cpp:921:5: warning: loop will run at most once (loop increment never executed) [-Wunreachable-code-loop-increment]
    for (const auto& arg : gArgs.GetUnsuitableSectionOnlyArgs()) {
    ^~~
1 warning generated.
```

> -Wunused-member-function
```bash
script/bitcoinconsensus.cpp:50:9: warning: unused member function 'GetType' [-Wunused-member-function]
    int GetType() const { return m_type; }
        ^
1 warning generated.

index/blockfilterindex.cpp:55:5: warning: unused member function 'DBHeightKey' [-Wunused-member-function]
    DBHeightKey() : height(0) {}
    ^
index/blockfilterindex.cpp:81:5: warning: unused function template 'Unserialize' [-Wunused-template]
    ADD_SERIALIZE_METHODS;
    ^
./serialize.h:198:10: note: expanded from macro 'ADD_SERIALIZE_METHODS'
    void Unserialize(Stream& s) {                                     \
         ^
2 warnings generated.

test/util_tests.cpp:1919:14: warning: unused member function 'operator=' [-Wunused-member-function]
    Tracker& operator=(Tracker&& t) noexcept
             ^
1 warning generated.
```

> -Wunreachable-code-return
```bash
leveldb/db/log_reader.cc:181:10: warning: 'return' will never be executed [-Wunreachable-code-return]
  return false;
         ^~~~~
1 warning generated.
```

--- a/configure.ac
+++ b/configure.ac
@@ -404,6 +404,10 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
   AX_CHECK_COMPILE_FLAG([-Wdate-time],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wdate-time"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wconditional-uninitialized],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wconditional-uninitialized"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wsign-compare],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wsign-compare"],,[[$CXXFLAG_WERROR]])
+  AX_CHECK_COMPILE_FLAG([-Wduplicated-branches],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wduplicated-branches"],,[[$CXXFLAG_WERROR]])
+  AX_CHECK_COMPILE_FLAG([-Wduplicated-cond],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wduplicated-cond"],,[[$CXXFLAG_WERROR]])
+  AX_CHECK_COMPILE_FLAG([-Wlogical-op],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wlogical-op"],,[[$CXXFLAG_WERROR]])
+  AX_CHECK_COMPILE_FLAG([-Woverloaded-virtual],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Woverloaded-virtual"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wsuggest-override],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wsuggest-override"],,[[$CXXFLAG_WERROR]],
                         [AC_LANG_SOURCE([[struct A { virtual void f(); }; struct B : A { void f() final; };]])])
   AX_CHECK_COMPILE_FLAG([-Wunreachable-code-loop-increment],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wunreachable-code-loop-increment"],,[[$CXXFLAG_WERROR]])

